File: scenarios/typescript/node/captioning/caption.ts
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

"use strict";

/* Notes:
- To install dependencies for this sample:
npm install microsoft-cognitiveservices-speech-sdk
- The Speech SDK on Windows requires the Microsoft Visual C++ Redistributable for Visual Studio 2019 on the system. See:
https://docs.microsoft.com/azure/cognitive-services/speech-service/speech-sdk?tabs=windows%2Cubuntu%2Cios-xcode%2Cmac-xcode%2Candroid-studio
*/

import { Buffer } from 'buffer';
import * as fs from "fs";
import * as sdk from "microsoft-cognitiveservices-speech-sdk";

// For definition of PullAudioInputStreamCallback interface, see:
// https://github.com/microsoft/cognitive-services-speech-sdk-js/blob/master/src/sdk/Audio/PullAudioInputStreamCallback.ts
class BinaryFileReader implements sdk.PullAudioInputStreamCallback {
    m_fd : number;

    constructor(audioFileName: string) {
        this.m_fd = fs.openSync(audioFileName, 'r');
// TODO1 If err, throw.
    }

    // See:
    // https://javascript.info/arraybuffer-binary-arrays
    read(dataBuffer: ArrayBuffer): number {
        var view = new Uint16Array(dataBuffer);
        var bytesRead = fs.readSync(this.m_fd, view);
        return bytesRead;
    }
    
    close(): void {
        fs.closeSync(this.m_fd);
    }
}

interface WavFileHeaderInfo {
    framerate : number;
    bitsPerSample : number;
    nChannels : number;
}

function ReadInt32(fd : number) : number {
    var buffer = Buffer.alloc(4);
    var bytesRead = fs.readSync(fd, buffer);
    // TODO1 Throw if bytesRead wrong.
    return buffer.readInt32LE();
}

function ReadUInt16(fd : number) : number {
    var buffer = Buffer.alloc(2);
    var bytesRead = fs.readSync(fd, buffer);
    // TODO1 Throw if bytesRead wrong.
    return buffer.readUInt16LE();
}

function ReadUInt32(fd : number) : number {
    var buffer = Buffer.alloc(4);
    var bytesRead = fs.readSync(fd, buffer);
    // TODO1 Throw if bytesRead wrong.
    return buffer.readUInt32LE();
}

function ReadString(fd : number, length : number) : string {
    var buffer = Buffer.alloc(length);
    var bytesRead = fs.readSync(fd, buffer);
    // TODO1 Throw if bytesRead wrong.
    return buffer.toString();
}

function ReadWavFileHeader(audioFileName : string) : WavFileHeaderInfo {
    var fd = fs.openSync(audioFileName, 'r');

    if(ReadString(fd, 4) != "RIFF") {
        throw "Invalid .wav file header.";
    }
    // File length
    ReadInt32(fd);
    if(ReadString(fd, 4) != "WAVE") {
        throw "Invalid .wav file header.";
    }
    if(ReadString(fd, 4) != "fmt ") {
        throw "Invalid .wav file header.";
    }
    // Format size
    // TODO1 Throw if >= 16
    ReadInt32(fd);
    // Format tag
    ReadUInt16(fd);
    var nChannels = ReadUInt16(fd);
    var framerate = ReadUInt32(fd);
    // Average bytes per second
    ReadUInt32(fd);
    // Block align
    ReadUInt16(fd);
    var bitsPerSample = ReadUInt16(fd);

    fs.closeSync(fd);

    return { framerate : framerate, bitsPerSample : bitsPerSample, nChannels : nChannels };
}

interface UserConfig {
    readonly profanityFilterRemoveEnabled : boolean;
    readonly profanityFilterMaskEnabled : boolean;
    readonly inputFile : string;
    readonly outputFile : string;
    readonly phraseList : string;
    readonly suppressOutputEnabled : boolean;
    readonly partialResultsEnabled : boolean;
    readonly stablePartialResultThreshold : string;
    readonly srtEnabled : boolean;
    readonly trueTextEnabled : boolean;
    readonly subscriptionKey : string;
    readonly region : string;
}

const newline = "\n";

function GetCmdOption(args : string[], option : string) : string
{
    const index = args.indexOf(option);
    if (index > -1 && index < args.length - 2)
    {
        // We found the option(for example, "-o"), so advance from that to the value(for example, "filename").
        return args[index + 1];
    }
    else {
        return null;
    }
}

function CmdOptionExists(args : string[], option : string) : boolean
{
    return args.includes(option);
}

function TimestampFromSpeechRecognitionResult(result : sdk.SpeechRecognitionResult, userConfig : UserConfig) : string
{
    /* Offset and duration are measured in 100-nanosecond increments. The Date constructor takes a value measured in milliseconds.
    100 nanoseconds is equal to a tick. There are 10,000 ticks in a millisecond.
    See:
    https://docs.microsoft.com/dotnet/api/system.timespan.ticks
    https://docs.microsoft.com/javascript/api/microsoft-cognitiveservices-speech-sdk/speechrecognitionresult
    */
    const ticksPerMillisecond = 10000;
    const startTime = new Date(result.offset/ticksPerMillisecond);
    const endTime = new Date(result.offset/ticksPerMillisecond + result.duration/ticksPerMillisecond);
    // Note We must use getUTC* methods, or the results are adjusted for our local time zone, which we don't want.
    const start_hours = startTime.getUTCHours().toString().padStart(2, '0');
    const start_minutes = startTime.getUTCMinutes().toString().padStart(2, '0');
    const start_seconds = startTime.getUTCSeconds().toString().padStart(2, '0');
    const start_milliseconds = startTime.getUTCMilliseconds().toString().padStart(3, '0');
    const end_hours = endTime.getUTCHours().toString().padStart(2, '0');
    const end_minutes = endTime.getUTCMinutes().toString().padStart(2, '0');
    const end_seconds = endTime.getUTCSeconds().toString().padStart(2, '0');
    const end_milliseconds = endTime.getUTCMilliseconds().toString().padStart(3, '0');    
    
    if (userConfig.srtEnabled)
    {
        // SRT format requires ',' as decimal separator rather than '.'.
        return `${start_hours}:${start_minutes}:${start_seconds},${start_milliseconds} --> ${end_hours}:${end_minutes}:${end_seconds},${end_milliseconds}`;
    }
    else
    {
        return `${start_hours}:${start_minutes}:${start_seconds}.${start_milliseconds} --> ${end_hours}:${end_minutes}:${end_seconds}.${end_milliseconds}`;
    }
}

function CaptionFromSpeechRecognitionResult(sequenceNumber : number, result : sdk.SpeechRecognitionResult, userConfig : UserConfig) : string
{
    var caption = "";
    if (!userConfig.partialResultsEnabled && userConfig.srtEnabled)
    {
        caption += `${sequenceNumber}${newline}`;
    }
    caption += `${TimestampFromSpeechRecognitionResult(result, userConfig)}${newline}`;
    caption += `${result.text}${newline}${newline}`
    return caption;
}

function WriteToConsole(text : string, userConfig : UserConfig) : void
{
    if (!userConfig.suppressOutputEnabled)
    {
        process.stdout.write(text);
    }
}

function WriteToConsoleOrFile(text : string, userConfig : UserConfig) : void
{
    WriteToConsole(text, userConfig);
    if (null !== userConfig.outputFile)
    {
        fs.appendFileSync(userConfig.outputFile, text);
    }
}

function Initialize(userConfig : UserConfig) : void
{
    if (null !== userConfig.outputFile && fs.existsSync(userConfig.outputFile))
    {
        fs.unlinkSync(userConfig.outputFile);
    }

    if (!userConfig.srtEnabled)
    {
        WriteToConsoleOrFile(`WEBVTT${newline}${newline}`, userConfig);
    }
}

// Note: We verified args.length >= 3 in the caller, main().
function UserConfigFromArgs(args : string[]) : UserConfig
{
    return {
        profanityFilterRemoveEnabled : CmdOptionExists(args, "-f"),
        profanityFilterMaskEnabled : CmdOptionExists(args, "-m"),
        inputFile : GetCmdOption(args, "-i"),
        outputFile : GetCmdOption(args, "-o"),
        phraseList : GetCmdOption(args, "-p"),
        suppressOutputEnabled : CmdOptionExists(args, "-q"),
        partialResultsEnabled : CmdOptionExists(args, "-u"),
        stablePartialResultThreshold : GetCmdOption(args, "-r"),
        srtEnabled : CmdOptionExists(args, "-s"),
        trueTextEnabled : CmdOptionExists(args, "-t"),
        subscriptionKey : args[args.length - 2],
        region : args[args.length - 1],
    };
}

function AudioConfigFromUserConfig(userConfig : UserConfig) : sdk.AudioConfig
{
    if (null !== userConfig.inputFile)
    {
        var header = ReadWavFileHeader(userConfig.inputFile);
        var format = sdk.AudioStreamFormat.getWaveFormatPCM(header.framerate, header.bitsPerSample, header.nChannels);
        var callback = new BinaryFileReader(userConfig.inputFile);
        var stream = sdk.AudioInputStream.createPullStream(callback, format);
        return sdk.AudioConfig.fromStreamInput(stream);
    }
    else
    {
        return sdk.AudioConfig.fromDefaultMicrophoneInput();
    }
}

function SpeechConfigFromUserConfig(userConfig : UserConfig) : sdk.SpeechConfig
{
    const speechConfig = sdk.SpeechConfig.fromSubscription(userConfig.subscriptionKey, userConfig.region);

    if (userConfig.profanityFilterRemoveEnabled)
    {
        speechConfig.setProfanity(sdk.ProfanityOption.Removed);
    }
    else if (userConfig.profanityFilterMaskEnabled)
    {
        speechConfig.setProfanity(sdk.ProfanityOption.Masked);
    }
    
    if (null !== userConfig.stablePartialResultThreshold)
    {
        speechConfig.setProperty("SpeechServiceResponse_StablePartialResultThreshold", userConfig.stablePartialResultThreshold);
    }
    
    speechConfig.setProperty("SpeechServiceResponse_PostProcessingOption", "TrueText");
    
    return speechConfig;
}

function SpeechRecognizerFromUserConfig(userConfig : UserConfig) : sdk.SpeechRecognizer
{
    const audioConfig = AudioConfigFromUserConfig(userConfig);
    const speechConfig = SpeechConfigFromUserConfig(userConfig);
    var speechRecognizer = new sdk.SpeechRecognizer(speechConfig, audioConfig);
    
    if (null !== userConfig.phraseList)
    {
        const grammar = sdk.PhraseListGrammar.fromRecognizer(speechRecognizer);
        grammar.addPhrase(userConfig.phraseList);
    }
    
    return speechRecognizer;
}

/* See:
https://docs.microsoft.com/azure/cognitive-services/speech-service/how-to-recognize-speech
*/
function RecognizeContinuous(speechRecognizer : sdk.SpeechRecognizer, userConfig : UserConfig) : void
{
    var sequenceNumber = 0;
    
    if (userConfig.partialResultsEnabled)
    {
        speechRecognizer.recognizing = function(s, e) {
            if (sdk.ResultReason.RecognizingSpeech == e.result.reason && e.result.text.length > 0)
            {
                // We don't show sequence numbers for partial results.
                WriteToConsole(CaptionFromSpeechRecognitionResult(0, e.result, userConfig), userConfig);
            }
            else if (sdk.ResultReason.NoMatch == e.result.reason)
            {
                WriteToConsole("NOMATCH: Speech could not be recognized.${newline}", userConfig);
            }
        };
    }

    speechRecognizer.recognized = function(s, e) {
        if (sdk.ResultReason.RecognizedSpeech == e.result.reason && e.result.text.length > 0)
        {
            sequenceNumber++;
            WriteToConsoleOrFile(CaptionFromSpeechRecognitionResult(sequenceNumber, e.result, userConfig), userConfig);
        }
        else if (sdk.ResultReason.NoMatch == e.result.reason)
        {
            // TODO1 String interpolation isn't working!
            WriteToConsole("NOMATCH: Speech could not be recognized.${newline}", userConfig);
        }
    };

    speechRecognizer.canceled =(s, e) => {
        if (sdk.CancellationReason.EndOfStream == e.reason)
        {
            WriteToConsole("End of stream reached.${newline}", userConfig);
        }
        else if (sdk.CancellationReason.Error == e.reason)
        {
            var error = "Encountered error.${newline}Error code: ${e.errorCode}${newline}Error details: ${e.errorDetails}${newline}";
        }
        else
        {
            var error = "Request was cancelled for an unrecognized reason: ${e.reason}.${newline}";
        }

        speechRecognizer.stopContinuousRecognitionAsync();
    };

    speechRecognizer.sessionStopped =(s, e) => {
        WriteToConsole("Session stopped.${newline}", userConfig);

        speechRecognizer.stopContinuousRecognitionAsync();
    };

    speechRecognizer.startContinuousRecognitionAsync();
}

function main(args : string[]) : void
{
    const usage = `Usage: node caption.js [-f] [-h] [-i file] [-m] [-o file] [-p phrases] [-q] [-r number] [-s] [-t] [-u] <subscriptionKey> <region>
            -f: Enable profanity filter (remove profanity). Overrides -m.
            -h: Show this help and stop.
            -i: Input audio file *file* (default input is from the microphone.)
            -m: Enable profanity filter (mask profanity). -f overrides this.
       -o file: Output to *file*.
    -p phrases: Add specified *phrases*.
                Example: Constoso;Jessie;Rehaan
            -q: Suppress console output (except errors).
     -r number: Set stable partial result threshold to *number*.
                Example: 3
            -s: Output captions in SRT format (default is WebVTT format.)
            -t: Enable TrueText.
            -u: Output partial results. These are always written to the console, never to an output file. -q overrides this.`;

    // Verify argc >= 3 (caption.exe, subscriptionKey, region)
    if (args.length < 3)
    {
        console.log("Too few arguments.");
        console.log(usage);
        return;
    }

    if (CmdOptionExists(args, "-h"))
    {
        console.log(usage);
    }
    else
    {
        const userConfig = UserConfigFromArgs(args);    
        Initialize(userConfig);    
        const audio_config = AudioConfigFromUserConfig(userConfig);
        const speechRecognizer = SpeechRecognizerFromUserConfig(userConfig);
        RecognizeContinuous(speechRecognizer, userConfig);
    }
}

main(process.argv);

File: scenarios/go/captioning/caption.go
//
// Copyright (c) Microsoft. All rights reserved.
// Licensed under the MIT license. See LICENSE.md file in the project root for full license information.
//

package main

// Notes:
// - Add Speech SDK dependency to your .mod file:
// go get -d github.com/Microsoft/cognitive-services-speech-sdk-go
// See also:
// https://docs.microsoft.com/azure/cognitive-services/speech-service/quickstarts/setup-platform
// - The Speech SDK on Windows requires the Microsoft Visual C++ Redistributable for Visual Studio 2019 on the system. See:
// https://docs.microsoft.com/azure/cognitive-services/speech-service/speech-sdk
// - Install gstreamer:
// https://docs.microsoft.com/azure/cognitive-services/speech-service/how-to-use-codec-compressed-audio-input-streams
// - In Windows, place the following files from the Speech SDK in the same folder as your .go file (if you are using `go run`) or compiled .exe:
//     - Microsoft.CognitiveServices.Speech.core.dll
//     - Microsoft.CognitiveServices.Speech.extension.audio.sys.dll

import(
    "bufio"
    "errors"
    "fmt"
    "github.com/Microsoft/cognitive-services-speech-sdk-go/common"
    "github.com/Microsoft/cognitive-services-speech-sdk-go/audio"
    "github.com/Microsoft/cognitive-services-speech-sdk-go/speech"    
    "log"
    "os"
    "strings"
    "time"
)

func GetCompressedAudioFormat(args []string) audio.AudioStreamContainerFormat {
    var value = GetCmdOption(args, "-c");
    if nil == value {
        return audio.ANY
    } else {
        switch strings.ToLower(*value) {
            case "alaw" : return audio.ALAW
            case "flac" : return audio.FLAC
            case "mp3" : return audio.MP3
            case "mulaw" : return audio.MULAW
            case "ogg_opus" : return audio.OGGOPUS
            default : return audio.ANY
        }
    }
}

func TimestampFromSpeechRecognitionResult(result speech.SpeechRecognitionResult) Timestamp {
    var startMilliseconds = time.Duration(result.Offset).Milliseconds()
    var endMilliseconds = time.Duration(result.Offset + result.Duration).Milliseconds()

    var startSeconds = int(startMilliseconds) / 1000
    var endSeconds = int(endMilliseconds) / 1000
   
    var startMinutes = startSeconds / 60
    var endMinutes = endSeconds / 60
    
    return Timestamp {
        startMilliseconds : int(startMilliseconds) % 1000,
        endMilliseconds : int(endMilliseconds) % 1000,
        startSeconds : startSeconds % 60,
        endSeconds : endSeconds % 60,
        startMinutes : startMinutes % 60,
        endMinutes : endMinutes % 60,
        startHours : startMinutes / 60,
        endHours : endMinutes / 60,
    }
}

func TimestampStringFromSpeechRecognitionResult(result speech.SpeechRecognitionResult, userConfig UserConfig) string {
    var time_format string
    var timestamp = TimestampFromSpeechRecognitionResult(result)
    if userConfig.useSubRipTextCaptionFormat {
        // SRT format requires ',' as decimal separator rather than '.'.
        time_format = "%02d:%02d:%02d,%03d"
    } else {
        time_format = "%02d:%02d:%02d.%03d"
    }
    return fmt.Sprintf(time_format + " --> " + time_format, timestamp.startHours, timestamp.startMinutes, timestamp.startSeconds, timestamp.startMilliseconds, timestamp.endHours, timestamp.endMinutes, timestamp.endSeconds, timestamp.endMilliseconds)
}

func CaptionFromSpeechRecognitionResult(sequenceNumber int, result speech.SpeechRecognitionResult, userConfig UserConfig) string {
    var caption strings.Builder
    if !userConfig.showRecognizingResults && userConfig.useSubRipTextCaptionFormat {
        caption.WriteString(fmt.Sprintf("%d%s", sequenceNumber, newline));
    }
    caption.WriteString(fmt.Sprintf("%s%s", TimestampStringFromSpeechRecognitionResult(result, userConfig), newline));
    caption.WriteString(fmt.Sprintf("%s%s%s", result.Text, newline, newline));
    return caption.String()
}

func Initialize(userConfig UserConfig) {
    if nil != userConfig.outputFile {
        _, err := os.Stat(*userConfig.outputFile)
        if nil == err {
            os.Remove(*userConfig.outputFile)
        } else if !errors.Is(err, os.ErrNotExist) {
            log.Fatal(err)
        }
    }

    if !userConfig.useSubRipTextCaptionFormat {
        WriteToConsoleOrFile(fmt.Sprintf("WEBVTT%s%s", newline, newline), userConfig)
    }
}

func UserConfigFromArgs(args []string) UserConfig {
    var userConfig = UserConfig {
        compressedAudioFormat : audio.ANY,
        removeProfanity : false,
        disableMaskingProfanity : false,
        inputFile : nil,
        outputFile : nil,
        suppressConsoleOutput : false,
        showRecognizingResults : false,
        stablePartialResultThreshold : nil,
        useSubRipTextCaptionFormat : false,
        subscriptionKey : "",
        region : "",
    }
    
// Verify argc >= 3 (caption.go, subscriptionKey, region)
    if len(args) < 3 {
        log.Fatal("Too few arguments.")
    }

    userConfig.subscriptionKey = args[len(args) - 2]
    userConfig.region = args[len(args) - 1]

    userConfig.compressedAudioFormat = GetCompressedAudioFormat(args)
    userConfig.removeProfanity = CmdOptionExists(args, "-f")
    userConfig.disableMaskingProfanity = CmdOptionExists(args, "-m")
    userConfig.inputFile = GetCmdOption(args, "-i")
    userConfig.outputFile = GetCmdOption(args, "-o")
    userConfig.showRecognizingResults = CmdOptionExists(args, "-u")
    userConfig.suppressConsoleOutput = CmdOptionExists(args, "-q");
    userConfig.stablePartialResultThreshold = GetCmdOption(args, "-r")
    userConfig.useSubRipTextCaptionFormat = CmdOptionExists(args, "-s")
    
    return userConfig
}

func AudioConfigFromUserConfig(userConfig UserConfig) *audio.AudioConfig {
    if nil != userConfig.inputFile {
        var format *audio.AudioStreamFormat
        var err error
        if strings.HasSuffix(*userConfig.inputFile, ".wav") {
            framerate, bitsPerSample, nChannels := ReadWavFileHeader(*userConfig.inputFile)
            format, err = audio.GetWaveFormatPCM(framerate, bitsPerSample, nChannels)
            if nil != err {
                log.Fatal(err)
            }
        } else {
            format, err = audio.GetCompressedFormat(userConfig.compressedAudioFormat)
            if nil != err {
                log.Fatal(err)
            }
            defer format.Close()
        }
        callback := GetBinaryFileReaderCallback(*userConfig.inputFile)
        stream, err := audio.CreatePullStreamFromFormat(callback, format)
        if nil != err {
            log.Fatal(err)
        }
        defer stream.Close()
        audioConfig, err := audio.NewAudioConfigFromStreamInput(stream)
        if nil != err {
            log.Fatal(err)
        }
        return audioConfig
    } else {
        result, err := audio.NewAudioConfigFromDefaultMicrophoneInput()
        if nil != err {
            log.Fatal(err)
        }
        return result
    }
}

func SpeechConfigFromUserConfig(userConfig UserConfig) *speech.SpeechConfig {
    var speechConfig, err = speech.NewSpeechConfigFromSubscription(userConfig.subscriptionKey, userConfig.region)
    if nil != err {
        log.Fatal(err)
    }

    if userConfig.removeProfanity {
        speechConfig.SetProfanity(common.Removed)
    } else if userConfig.disableMaskingProfanity {
        speechConfig.SetProfanity(common.Masked)
    }
    
    if nil != userConfig.stablePartialResultThreshold {
        speechConfig.SetProperty(common.SpeechServiceResponseStablePartialResultThreshold, *userConfig.stablePartialResultThreshold)
    }
    
    speechConfig.SetProperty(common.SpeechServiceResponsePostProcessingOption, "TrueText")

    return speechConfig
}

func SpeechRecognizerFromUserConfig(userConfig UserConfig) *speech.SpeechRecognizer {
    var audioConfig = AudioConfigFromUserConfig(userConfig)
    defer audioConfig.Close()
    var speechConfig = SpeechConfigFromUserConfig(userConfig)
    defer speechConfig.Close()
    var speechRecognizer, err = speech.NewSpeechRecognizerFromConfig(speechConfig, audioConfig)
    if nil != err {
        log.Fatal(err)
    }
    
    return speechRecognizer
}

func GetRecognizingHandler(userConfig UserConfig) func(speech.SpeechRecognitionEventArgs) {
    return func(e speech.SpeechRecognitionEventArgs) {
        defer e.Close()
        if common.RecognizingSpeech == e.Result.Reason && len(e.Result.Text) > 0 {
            // We don't show sequence numbers for partial results.
            WriteToConsole(CaptionFromSpeechRecognitionResult(0, e.Result, userConfig), userConfig);
        } else if common.NoMatch == e.Result.Reason {
            WriteToConsole("NOMATCH: Speech could not be recognized.\n", userConfig);
        }
    }
}

func GetRecognizedHandler(userConfig UserConfig) func(speech.SpeechRecognitionEventArgs) {
    return func(e speech.SpeechRecognitionEventArgs) {
        defer e.Close()
        var sequenceNumber int = 0
        if common.RecognizedSpeech == e.Result.Reason && len(e.Result.Text) > 0 {
            sequenceNumber++
            WriteToConsoleOrFile(CaptionFromSpeechRecognitionResult(sequenceNumber, e.Result, userConfig), userConfig);
        } else if common.NoMatch == e.Result.Reason {
            WriteToConsole(fmt.Sprintf("NOMATCH: Speech could not be recognized.%s", newline), userConfig)
        }
    }
}

func GetSessionStoppedHandler(userConfig UserConfig) func(speech.SessionEventArgs) {
    return func(e speech.SessionEventArgs) {
        defer e.Close()
        WriteToConsole(fmt.Sprintf("Session stopped.%s", newline), userConfig)
    }
}

func GetCanceledHandler(userConfig UserConfig) func(speech.SpeechRecognitionCanceledEventArgs) {
    return func(e speech.SpeechRecognitionCanceledEventArgs) {
        defer e.Close()
        if common.EndOfStream == e.Reason {
            WriteToConsole(fmt.Sprintf("End of stream reached.%s", newline), userConfig)
        } else if common.CancelledByUser == e.Reason {
            WriteToConsole(fmt.Sprintf("User canceled request.%s", newline), userConfig)
        } else if common.Error == e.Reason {
            WriteToConsole(fmt.Sprintf("Encountered error.%sError code: %d%sError details: %s%s", newline, int(e.ErrorCode), newline, e.ErrorDetails, newline), userConfig)
        } else {
            WriteToConsole(fmt.Sprintf("Request was cancelled for an unrecognized reason: %d.%s", int(e.Reason), newline), userConfig)
        }
    }
}

func recognizeContinuous(speechRecognizer *speech.SpeechRecognizer, userConfig UserConfig) {
    if userConfig.showRecognizingResults {
        speechRecognizer.Recognizing(GetRecognizingHandler(userConfig))
    }
    speechRecognizer.Recognized(GetRecognizedHandler(userConfig))
    speechRecognizer.SessionStopped(GetSessionStoppedHandler(userConfig))
    speechRecognizer.Canceled(GetCanceledHandler(userConfig))
    
    speechRecognizer.StartContinuousRecognitionAsync()
    defer speechRecognizer.StopContinuousRecognitionAsync()
    // Keep function from exiting before we are done.
    bufio.NewReader(os.Stdin).ReadBytes('\n')
}

func main() {
    var usage string = `Usage: go run caption.go [-c ALAW|ANY|FLAC|MP3|MULAW|OGG_OPUS] [-f] [-h] [-i file] [-m] [-o file] [-q] [-s] [-t number] [-u] <subscriptionKey> <region>
       -c format: Use compressed audio format.
                  Valid values: ALAW, ANY, FLAC, MP3, MULAW, OGG_OPUS.
                  Default value: ANY.
              -f: Remove profanity (default behavior is to mask profanity). Overrides -m.
              -h: Show this help and stop.
              -i: Input audio file *file* (default input is from the microphone.)
              -m: Disable masking profanity (default behavior). -f overrides this.
         -o file: Output to *file*.
              -q: Suppress console output (except errors).
              -s: Output captions in SRT format (default is WebVTT format.)
       -t number: Set stable partial result threshold to *number*.
                  Example: 3
              -u: Output partial results. These are always written to the console, never to an output file. -q overrides this.`

    if CmdOptionExists(os.Args, "-h") {
        fmt.Println(usage);
    } else {
        var userConfig = UserConfigFromArgs(os.Args)
        Initialize(userConfig)
        var speechRecognizer = SpeechRecognizerFromUserConfig(userConfig)
        defer speechRecognizer.Close()
        recognizeContinuous(speechRecognizer, userConfig)
    }
}
